onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/reduction_op_builder.cc

```diff
@@ -0,0 +1,234 @@
+      Shape axes_dimen = {static_cast<uint32_t>(axes.size())};
+      const OperandType axes_operand_type(Type::TENSOR_INT32, axes_dimen);
+      ORT_RETURN_IF_ERROR(model_builder.AddOperandFromPersistMemoryBuffer(axes_name, axes.data(), axes_operand_type));
+
+      input_indices.push_back(operand_indices.at(axes_name));  // axes
+
+      int32_t input_size = static_cast<int32_t>(input_shape.size());
+      std::unordered_set<int32_t> axes_to_be_reduced;
+
+      for (const auto& axis : axes) {
+        axes_to_be_reduced.insert(axis);
+      }
+
+      // Make output dimensions
+      InlinedVector<uint32_t> output_dimen;
+      if (keepdims == 1) {
+        output_dimen.reserve(input_size);
+      } else {
+        output_dimen.reserve(input_size - axes_to_be_reduced.size());
+      }
+
+      for (int32_t i = 0; i < input_size; i++) {
+        if (!Contains(axes_to_be_reduced, i)) {
+          output_dimen.push_back(input_shape[i]);
+        } else {
+          if (keepdims == 1) {
+            output_dimen.push_back(1);
+          }
+        }
+      }
+
+      // In case of a tensor has all 1's in dimension such as {1,1,1,1} and gets all reduced,
+      // NNAPI requires the output shape to be {1}. (otherwise NNAPI will treat it as dynamic shape.)
+      if (output_dimen.empty())
+        output_dimen.push_back(1);
+
+      shaper.AddShape(output, output_dimen);
+
+      ADD_SCALAR_OPERAND(model_builder, input_indices, keepdims);
+
+      const OperandType output_operand_type(operand_types.at(inputs[0].node_arg.Name()).type, output_dimen);
+      ORT_RETURN_IF_ERROR(model_builder.AddOperation(op_code, input_indices,
+                                                     {output}, {output_operand_type}));
+    } else {
+      // For case that `axes` is empty, act as an Identity op
+      const OperandType output_operand_type(operand_types.at(inputs[0].node_arg.Name()).type, input_shape);
+      model_builder.RegisterOperand(output, operand_indices.at(inputs[0].node_arg.Name()), output_operand_type);
+    }
+  }
+
+  return Status::OK();
+}
+
+// Operator support related
+
+int32_t ReductionOpBuilder::GetMinSupportedNNAPIFeatureLevel(
+    const NodeUnit& node_unit, const OpSupportCheckParams& /* params */) const {
+  const auto& op(node_unit.OpType());
+  if (op == "ReduceMean") {
+    return ANEURALNETWORKS_FEATURE_LEVEL_2;
+  }
+
+  return ANEURALNETWORKS_FEATURE_LEVEL_3;
+}
+
+bool ReductionOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializers, const NodeUnit& node_unit,
+                                           const OpSupportCheckParams& /* params */) const {
+  const auto& inputs = node_unit.Inputs();
+  const auto& op(node_unit.OpType());
+
+  Shape input_shape;
+  if (!GetShape(inputs[0].node_arg, input_shape))
+    return false;
+
+  if (input_shape.size() > 4 || input_shape.empty()) {
+    LOGS_DEFAULT(VERBOSE) << "NNAPI reduction ops only support 1-4d shape, input is "
```

Comment:
checked for most other reduction ops in NNAPI doc, e,g, ANEURALNETWORKS_REDUCE_MIN/MAX/PROD/SUM,...etc. 

supported tensor rank is all set as "up to 4".  

For instance, ANEURALNETWORKS_REDUCE_MAX: https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0a257a84d4cfa9f6eef0997ddd591e9584